File: src/src/Aquarium.cpp
#include "Aquarium.hpp"
#include "Fish.hpp"
#include "ObjekMati.hpp"
#include "Pet.hpp"
#include <bits/stdc++.h>
using namespace std;

#define DEBUG_START cout<<"DEBUG_START\n";
#define DEBUG_END   cout<<"DEBUG_END\n";

//ctor
Aquarium::Aquarium():LENGTH(100),WIDTH(100){
}
//ctor with set the LENGTH and WIDTH
Aquarium::Aquarium(int length,int width):LENGTH(length),WIDTH(width){
}

//ctor with LENGTH and WIDTH and defined list
Aquarium::Aquarium(int length,int width,const LinkedList<ObjekMati*>& ListObjekMati,
 const LinkedList<Fish*>& ListIkan, const LinkedList<Pet*>& ListPet):LENGTH(length),WIDTH(width){
     List_Objek_Mati = ListObjekMati;
     List_Ikan = ListIkan;
     List_Pet = ListPet;
}
//dtor
Aquarium::~Aquarium(){
    delete[] &List_Objek_Mati;
    delete[] &List_Ikan;
    delete[] &List_Pet;
}

//add an object to list
void Aquarium::AddObject(Fish* fish){
    List_Ikan.add(fish);
}
void Aquarium::AddObject(ObjekMati* objekMati){
    List_Objek_Mati.add(objekMati);
}
void Aquarium::AddObject(Pet* pet){
    List_Pet.add(pet);
}

//remove an object from list
void Aquarium::RemoveObject(Fish* fish){
    List_Ikan.remove(fish);
}

void Aquarium::RemoveObject(ObjekMati* objekMati){
    List_Objek_Mati.remove(objekMati);
}
void Aquarium::RemoveObject(Pet* pet){
    List_Pet.remove(pet);
}

//Make all of the instances in the do something
void Aquarium::Action(){
    for(int i=0;i<List_Ikan.totalElmt();i++){
        Fish *temp = List_Ikan.get(i);
        if(temp->getStarving()<=0){
            this->RemoveObject(List_Ikan.get(i));
        }
        else{
            int count=0,countguppy=0;
            for(int j=0;j<List_Objek_Mati.totalElmt();j++){
                if(List_Objek_Mati.get(j)->getJenis()=="Makanan Ikan"){
                    count++;
                }
            }
            for(int j=0;j<List_Ikan.totalElmt();j++){
                if(List_Ikan.get(j)->getJenis()=="Guppy"){
                    countguppy++;
                }
            }
            if(temp->getJenis()=="Guppy"){
                if(temp->getStarvationPeriod() > 0 || count==0){
                    temp->setStarvationPeriod(temp->getStarvationPeriod()-1);
                    if(temp->getMoveTime()<=0){
                        //srand(time(NULL));
                        int randoms = (rand() % 40);
                        cout<<randoms<<" haha\n";
                        int directions = (rand()%2);
                        if(directions){
                            randoms*=-1;
                        }
                        temp->setDegree((temp->getDegree()+randoms)%360);
                        temp->setMoveTime(temp->getMAXMOVE());
                    }
                    temp->swim(temp->getDegree(),temp->getSpeed());
                    temp->setStarving(temp->getStarving()-1);
                }
                else{
                    temp->eat(*this);
                }
            }
            else{
                if(temp->getStarvationPeriod() > 0 || countguppy==0){
                    temp->setStarvationPeriod(temp->getStarvationPeriod()-1);
                    if(temp->getMoveTime()<=0){
                        //srand(time(NULL));
                        int randoms = (rand() % 360);

                        int directions = (rand()%2);
                        if(directions){
                            randoms*=-1;
                        }
                        cout<<randoms<<" haha\n";
                        temp->setDegree((temp->getDegree()+randoms)%360);
                        temp->setMoveTime(temp->getMAXMOVE());
                    }
                    temp->swim(temp->getDegree(),temp->getSpeed());
                    temp->setStarving(temp->getStarving()-1);
                }
                else{
                    temp->eat(*this);
                }
            }
            temp->setMoveTime(temp->getMoveTime()-1);

            //produce
            if(temp->getJenis() == "Guppy"){
                if(temp->getProduceTime() < 0){
                    temp->produce(*this);
                    temp->setProduceTime(temp->getMaxProduceTime());
                }
                //cout<<temp->getProduceTime()<<" ampun\n";
                temp->setProduceTime(temp->getProduceTime()-1);
            }
        }
    }
    for(int i=0;i<List_Pet.totalElmt();i++){
        List_Pet.get(i)->eat(*this);
        //cout<<"EAT TELAH DIJALANKAN\n";
    }
    for(int i=0;i<List_Objek_Mati.totalElmt();i++){
        List_Objek_Mati.get(i)->turun(*this);
    }
}

//Check whether a point is an edge or not in a radius
//RETURN :
//0 Not Edge 1 UP      2 RIGHT     3 DOWN      4 LEFT
int Aquarium::IsEdge(Point& P, double radius){
    if(abs(P.getOrdinat() - WIDTH) < radius){
        return 2;
    }
    else if(P.getOrdinat() < radius){
        return 4;
    }
    else if(abs(P.getAbsis() - LENGTH) < radius){
        return 1;
    }
    else if(P.getAbsis() < radius){
        return 3;
    }
    else{
        return 0;
    }
}
LinkedList<ObjekMati*>& Aquarium::getListObjekMati(){return List_Objek_Mati;}
LinkedList<Fish*>& Aquarium::getListIkan(){return List_Ikan;}
LinkedList<Pet*>& Aquarium::getListPet(){return List_Pet;}
int Aquarium::getLength(){return LENGTH;}
int Aquarium::getWidth(){return WIDTH;}
int Aquarium::getTime(){return time;}

File: src/src/Aquarium.java
//CARA COMPILE
//javac Aquarium.java Fish.java Guppy.java Piranha.java
// ObjekMati.java Koin.java MakananIkan.java Pet.java Point.java
import java.lang.Math;

/**
 * Tugas Besar 2 Pemrograman Berorientasi Objek.
 * @author K03-L
 * @version 1.0
 * @since 1.0
 */
public class Aquarium {

  private LinkedList<ObjekMati> listObjekMati = new LinkedList<ObjekMati>();
  private LinkedList<Fish> listIkan = new LinkedList<Fish>();
  private LinkedList<Pet> listPet = new LinkedList<Pet>();
  private int time;
  private int dasar;
  private final int length;
  private final int width;

  /**
   * Constructor Aquarium, mengassign nilai length dan width menjadi 100.
   */
  public Aquarium() {
    length = 100;
    width = 100;
  }

  /**
   * Constructor Aquarium, mengassign nilai length dan width sesuai masukan.
   * @param length ukuran length (ATAS-BAWAH) akuarium
   * @param width ukuran width (KIRI-KANAN) akuarium
   */
  public Aquarium(int length, int width) {
    this.length = length;
    this.width = width;
  }

  /**
   * Constructor Aquarium dengan mengassign nilai length,
   * width, dan list benda-benda yang ada di aquarium.
   * @param length ukuran length (ATAS-BAWAH) akuarium
   * @param width ukuran width (KIRI-KANAN) akuarium
   * @param listObjekMati list dari objek mati yang ingin dimasukkan ke akuarium
   * @param listIkan list dari ikan yang ingin dimasukkan ke akuarium
   * @param listPet list dair pet yang ingin dimasukkan ke akuarium
   */
  public Aquarium(int length, int width, LinkedList<ObjekMati> listObjekMati,
                  LinkedList<Fish> listIkan, LinkedList<Pet> listPet) {
    this.length = length;
    this.width = width;
    this.listObjekMati = listObjekMati;
    this.listIkan = listIkan;
    this.listPet = listPet;
  }

  /**
   * Menambahkan sebuah objek bertipe Fish ke dalam listIkan.
   * @param fish ikan yang ingin dimasukkan ke dalam list ikan
   */
  public void addObject(Fish fish) {
    System.out.println(fish.getPosisi().getAbsis());
    listIkan.add(fish);
  }

  /**
   * Menambahkan sebuah objek bertipe objekMati ke dalam listObjekMati.
   * @param objekMati objek mati yang ingin dimasukkan ke dalam list  objek mati
   */
  public void addObject(ObjekMati objekMati) {
    listObjekMati.add(objekMati);
  }

  /**
   * Menambahkan sebuah objek bertipe Pet ke dalam listPet.
   * @param pet pet yang ingin dimasukkan ke dalam list pet
   */
  public void addObject(Pet pet) {
    listPet.add(pet);
  }

  /**
   * Menghapus suatu ikan pada list ikan.
   * @param fish ikan yang ingin dihapus dari list ikan
   */
  public void removeObject(Fish fish) {
    listIkan.remove(fish);
  }

  /**
   * Menghapus suatu objekMati pada liatObjekMati.
   * @param objekMati objek mati yang ingin dihapus list objek mati.
   */
  public void removeObject(ObjekMati objekMati) {
    listObjekMati.remove(objekMati);
  }

  /**
   * Menghapus suatu pet dari listPet.
   * @param pet pet yang ingin dihapus dari list pet
   */
  public void removeObject(Pet pet) {
    listPet.remove(pet);
  }

  /**
   * Menghitung Banyaknya Guppy.
   */
  public int countGuppy() {
    int countguppy = 0;
    for (int j = 0; j < listIkan.totalElmt(); j++) {
      if (listIkan.get(j).getJenis().equals("Guppy")) {
        countguppy++;
      }
    }
    return countguppy;
  }

  /**
   * Menghitung Banyaknya Makanan ikan.
   */
  public int countFood() {
    int countfood = 0; 
    for (int j = 0; j < listObjekMati.totalElmt(); j++) {
      if (listObjekMati.get(j).getJenis().equals("Makanan Ikan")) {
        countfood++;
      }
    }
    return countfood;
  }

  /**
   * Mengendalikan kehidupan ikan.
   */
  public void controlFish() {
    for (int i = 0; i < listIkan.totalElmt(); i++) {
      Fish tempoint = listIkan.get(i);
      if (tempoint.getStarving() <= 0) {
        removeObject(listIkan.get(i));
      } else {
        if (tempoint.getJenis().equals("Guppy")) {
          if (tempoint.getStarvationPeriod() > 0 || countFood() == 0) {
            tempoint.trySwim();
          } else {
            tempoint.hunt(this);
          }
          tempoint.tryProduce(this);
        } else {  //Piranha
          if (tempoint.getStarvationPeriod() > 0 || countGuppy() == 0) {
            tempoint.trySwim();
          } else {
            tempoint.hunt(this);
          }
        }
        tempoint.setMoveTime(tempoint.getMoveTime() - 1);
      }
    }
  }

  /**
   * Mengendalikan kehidupan pet.
   */
  public void controlPet() {
    for (int i = 0; i < listPet.totalElmt(); i++) {
      listPet.get(i).life(this);
    }
  }

  /**
   * Mengendalikan kehidupan benda mati.
   */
  public void controlDeadObject() {
    for (int i = 0; i < listObjekMati.totalElmt(); i++) {
      listObjekMati.get(i).turun(this);
    }
  }

  /**
   * Mengendalikan kehidupan dalam akuarium.
   */
  public void action() {
    controlFish();
    controlPet();
    controlDeadObject();
  }

  /**
   * Mengecek apakah suatu titik merupakan ujung dari akuarium
   * dalam radius tertentu.
   * @param point sebuah point yang ingin dicek
   * @param radius jarak ketelitian P terhadap edge
   * @return 0 BUKAN UJUNG 1 ATAS 2 KANAN 3 BAWAH 4 KIRI
   */
  public int isEdge(Point point, double radius) {
    if (Math.abs(point.getOrdinat() - width) < radius) {
      return 2;
    } else if (point.getOrdinat() < radius) {
      return 4;
    } else if (Math.abs(point.getAbsis() - length) < radius) {
      return 1;
    } else if (point.getAbsis() < radius) {
      return 3;
    } else {
      return 0;
    }
  }

  public LinkedList<ObjekMati> getListObjekMati() {
    return listObjekMati;
  }

  public LinkedList<Fish> getListIkan() {
    return listIkan;
  }

  public LinkedList<Pet> getListPet() {
    return listPet;
  }

  public int getLength() {
    return length;
  }

  public int getWidth() {
    return width;
  }

  public int getTime() {
    return time;
  }

  public int getDasar() {
    return 6 * (length / 7);
  }
}
